File: frontend/server/routes/snippets.js
/**
 * Snippets API Routes
 * 代码片段管理 API
 */

import express from 'express';
import { db } from '../database/db.js';

const router = express.Router();

// ============================================
// GET /api/snippets - 获取代码片段列表
// ============================================
router.get('/', (req, res) => {
  try {
    const { search, category, language, limit = 50, offset = 0 } = req.query;
    
    let query = 'SELECT * FROM snippets WHERE 1=1';
    const params = [];

    if (search) {
      query += ` AND (title LIKE ? OR description LIKE ? OR tags LIKE ?)`;
      params.push(`%${search}%`, `%${search}%`, `%${search}%`);
    }

    if (category) {
      query += ` AND category = ?`;
      params.push(category);
    }

    if (language) {
      query += ` AND language = ?`;
      params.push(language);
    }

    query += ` ORDER BY updated_at DESC LIMIT ? OFFSET ?`;
    params.push(parseInt(limit), parseInt(offset));

    const stmt = db.prepare(query);
    const result = stmt.all(...params);
    
    res.json({
      success: true,
      snippets: result,
      total: result.length
    });
  } catch (error) {
    console.error('[Snippets API] 获取片段列表失败:', error);
    res.status(500).json({
      success: false,
      error: '获取片段列表失败'
    });
  }
});

// ============================================
// GET /api/snippets/popular - 获取热门片段
// ============================================
router.get('/popular', (req, res) => {
  try {
    const { limit = 10 } = req.query;
    
    const query = `
      SELECT s.*, COUNT(u.usage_id) as usage_count
      FROM snippets s
      LEFT JOIN snippet_usage u ON s.snippet_id = u.snippet_id
      GROUP BY s.snippet_id
      ORDER BY usage_count DESC, s.created_at DESC
      LIMIT ?
    `;
    
    const stmt = db.prepare(query);
    const result = stmt.all(parseInt(limit));
    
    res.json({
      success: true,
      snippets: result
    });
  } catch (error) {
    console.error('[Snippets API] 获取热门片段失败:', error);
    res.status(500).json({
      success: false,
      error: '获取热门片段失败'
    });
  }
});

// ============================================
// GET /api/snippets/recent - 获取最近片段
// ============================================
router.get('/recent', (req, res) => {
  try {
    const { limit = 10 } = req.query;
    
    const query = `
      SELECT s.*, MAX(u.used_at) as last_used_at
      FROM snippets s
      LEFT JOIN snippet_usage u ON s.snippet_id = u.snippet_id
      GROUP BY s.snippet_id
      ORDER BY last_used_at DESC NULLS LAST, s.updated_at DESC
      LIMIT ?
    `;
    
    const stmt = db.prepare(query);
    const result = stmt.all(parseInt(limit));
    
    res.json({
      success: true,
      snippets: result
    });
  } catch (error) {
    console.error('[Snippets API] 获取最近片段失败:', error);
    res.status(500).json({
      success: false,
      error: '获取最近片段失败'
    });
  }
});

// ============================================
// GET /api/snippets/categories - 获取分类列表（公开）
// ============================================
router.get('/categories', (req, res) => {
  try {
    const query = `
      SELECT DISTINCT category, COUNT(*) as count
      FROM snippets
      GROUP BY category
      ORDER BY count DESC, category
    `;
    
    const stmt = db.prepare(query);
    const result = stmt.all();
    
    res.json({
      success: true,
      categories: result.map(row => row.category)
    });
  } catch (error) {
    console.error('[Snippets API] 获取分类列表失败:', error);
    res.status(500).json({
      success: false,
      error: '获取分类列表失败'
    });
  }
});

// ============================================
// GET /api/snippets/tags - 获取标签列表（公开）
// ============================================
router.get('/tags', (req, res) => {
  try {
    // SQLite 不支持 unnest，使用不同的方法
    const query = `
      SELECT DISTINCT json_each.value as tag
      FROM snippets, json_each(snippets.tags)
      WHERE length(snippets.tags) > 0
      ORDER BY tag
    `;
    
    const stmt = db.prepare(query);
    const result = stmt.all();
    
    res.json({
      success: true,
      tags: result.map(row => row.tag)
    });
  } catch (error) {
    console.error('[Snippets API] 获取标签列表失败:', error);
    res.status(500).json({
      success: false,
      error: '获取标签列表失败'
    });
  }
});

// ============================================
// GET /api/snippets/:id - 获取单个片段
// ============================================
router.get('/:id', (req, res) => {
  try {
    const { id } = req.params;
    
    const query = 'SELECT * FROM snippets WHERE snippet_id = ?';
    const stmt = db.prepare(query);
    const result = stmt.get(id);
    
    if (!result) {
      return res.status(404).json({
        success: false,
        error: '片段不存在'
      });
    }
    
    // 记录使用
    const insertUsage = db.prepare('INSERT INTO snippet_usage (snippet_id, used_at) VALUES (?, datetime(\'now\'))');
    insertUsage.run(id);
    
    res.json({
      success: true,
      snippet: result
    });
  } catch (error) {
    console.error('[Snippets API] 获取片段失败:', error);
    res.status(500).json({
      success: false,
      error: '获取片段失败'
    });
  }
});

// ============================================
// POST /api/snippets - 创建新片段
// ============================================
router.post('/', (req, res) => {
  try {
    const { title, code, language, category, description, tags } = req.body;
    
    if (!title || !code) {
      return res.status(400).json({
        success: false,
        error: '标题和代码内容不能为空'
      });
    }
    
    const query = `
      INSERT INTO snippets (title, code, language, category, description, tags, created_at, updated_at)
      VALUES (?, ?, ?, ?, ?, ?, datetime('now'), datetime('now'))
      RETURNING *
    `;
    
    const stmt = db.prepare(query);
    const result = stmt.run(title, code, language || 'javascript', category || '通用', description || '', JSON.stringify(tags || []));
    
    // 获取插入的数据
    const insertedSnippet = db.prepare('SELECT * FROM snippets WHERE snippet_id = ?').get(result.lastInsertRowid);
    
    res.json({
      success: true,
      snippet: insertedSnippet
    });
  } catch (error) {
    console.error('[Snippets API] 创建片段失败:', error);
    res.status(500).json({
      success: false,
      error: '创建片段失败'
    });
  }
});

// ============================================
// PUT /api/snippets/:id - 更新片段
// ============================================
router.put('/:id', (req, res) => {
  try {
    const { id } = req.params;
    const { title, code, language, category, description, tags } = req.body;
    
    const query = `
      UPDATE snippets
      SET title = ?, code = ?, language = ?, category = ?, description = ?, tags = ?, updated_at = datetime('now')
      WHERE snippet_id = ?
      RETURNING *
    `;
    
    const stmt = db.prepare(query);
    const result = stmt.run(title, code, language, category, description, JSON.stringify(tags), id);
    
    if (result.changes === 0) {
      return res.status(404).json({
        success: false,
        error: '片段不存在'
      });
    }
    
    const updatedSnippet = db.prepare('SELECT * FROM snippets WHERE snippet_id = ?').get(id);
    
    res.json({
      success: true,
      snippet: updatedSnippet
    });
  } catch (error) {
    console.error('[Snippets API] 更新片段失败:', error);
    res.status(500).json({
      success: false,
      error: '更新片段失败'
    });
  }
});

// ============================================
// DELETE /api/snippets/:id - 删除片段
// ============================================
router.delete('/:id', (req, res) => {
  try {
    const { id } = req.params;
    
    const query = 'DELETE FROM snippets WHERE snippet_id = ?';
    const stmt = db.prepare(query);
    const result = stmt.run(id);
    
    if (result.changes === 0) {
      return res.status(404).json({
        success: false,
        error: '片段不存在'
      });
    }
    
    res.json({
      success: true,
      message: '片段删除成功'
    });
  } catch (error) {
    console.error('[Snippets API] 删除片段失败:', error);
    res.status(500).json({
      success: false,
      error: '删除片段失败'
    });
  }
});

export default router;

File: backend/app/routers/snippets.py
"""
Snippets Router
代码片段管理相关的 API 路由
"""

import json
import logging
from typing import List, Optional, Dict, Any
from fastapi import APIRouter, Query
from fastapi.responses import JSONResponse
from pydantic import BaseModel

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/snippets", tags=["Snippets"])


# ============================================
# 数据库连接函数
# ============================================

def get_db_connection():
    """获取数据库连接"""
    import sqlite3
    db_path = "storage/frontend/snippets.db"
    return sqlite3.connect(db_path)


# ============================================
# 数据模型
# ============================================

class SnippetCreate(BaseModel):
    title: str
    code: str
    language: str = "javascript"
    category: str = "通用"
    description: str = ""
    tags: List[str] = []


class SnippetUpdate(BaseModel):
    title: Optional[str] = None
    code: Optional[str] = None
    language: Optional[str] = None
    category: Optional[str] = None
    description: Optional[str] = None
    tags: Optional[List[str]] = None


# ============================================
# Snippets 端点
# ============================================

@router.get("")
async def get_snippets(
    search: Optional[str] = None,
    category: Optional[str] = None,
    language: Optional[str] = None,
    favorite_only: bool = False
):
    """获取代码片段列表"""
    try:
        conn = get_db_connection()
        cursor = conn.cursor()

        query = "SELECT * FROM snippets WHERE 1=1"
        params = []

        if search:
            query += " AND (title LIKE ? OR description LIKE ? OR code LIKE ?)"
            params.extend([f"%{search}%", f"%{search}%", f"%{search}%"])

        if category:
            query += " AND category = ?"
            params.append(category)

        if language:
            query += " AND language = ?"
            params.append(language)

        if favorite_only:
            query += " AND is_favorite = 1"

        query += " ORDER BY updated_at DESC"

        cursor.execute(query, params)
        rows = cursor.fetchall()

        snippets = []
        for row in rows:
            snippet = dict(row)
            snippet['tags'] = json.loads(snippet['tags']) if snippet['tags'] else []
            snippets.append(snippet)

        # 获取分类和标签
        categories = [row[0] for row in cursor.execute("SELECT DISTINCT category FROM snippets ORDER BY category")]
        tags = set()
        for snippet in snippets:
            tags.update(snippet['tags'])

        conn.close()

        return JSONResponse({
            "snippets": snippets,
            "categories": categories,
            "tags": list(tags)
        })
    except Exception as e:
        logger.exception(f"获取代码片段失败: {e}")
        return JSONResponse({"error": str(e)}, status_code=500)


@router.post("")
async def create_snippet(snippet: SnippetCreate):
    """创建代码片段"""
    try:
        conn = get_db_connection()
        cursor = conn.cursor()

        cursor.execute('''
            INSERT INTO snippets (title, code, language, category, description, tags)
            VALUES (?, ?, ?, ?, ?, ?)
        ''', (
            snippet.title,
            snippet.code,
            snippet.language,
            snippet.category,
            snippet.description,
            json.dumps(snippet.tags)
        ))

        snippet_id = cursor.lastrowid
        conn.commit()
        conn.close()

        return JSONResponse({"id": snippet_id, "message": "代码片段创建成功"})
    except Exception as e:
        logger.exception(f"创建代码片段失败: {e}")
        return JSONResponse({"error": str(e)}, status_code=500)


@router.get("/categories")
async def get_snippet_categories():
    """获取代码片段分类"""
    try:
        conn = get_db_connection()
        cursor = conn.cursor()

        cursor.execute("SELECT DISTINCT category FROM snippets ORDER BY category")
        categories = [row[0] for row in cursor.fetchall()]

        conn.close()

        return JSONResponse({"categories": categories})
    except Exception as e:
        logger.exception(f"获取代码片段分类失败: {e}")
        return JSONResponse({"error": str(e)}, status_code=500)


@router.get("/tags")
async def get_snippet_tags():
    """获取代码片段标签"""
    try:
        conn = get_db_connection()
        cursor = conn.cursor()

        cursor.execute("SELECT tags FROM snippets")
        all_tags = set()
        for row in cursor.fetchall():
            if row[0]:
                tags = json.loads(row[0])
                all_tags.update(tags)

        conn.close()

        return JSONResponse({"tags": list(all_tags)})
    except Exception as e:
        logger.exception(f"获取代码片段标签失败: {e}")
        return JSONResponse({"error": str(e)}, status_code=500)


@router.get("/{snippet_id}")
async def get_snippet(snippet_id: int):
    """获取单个代码片段"""
    try:
        conn = get_db_connection()
        cursor = conn.cursor()

        cursor.execute("SELECT * FROM snippets WHERE id = ?", (snippet_id,))
        row = cursor.fetchone()

        if not row:
            conn.close()
            return JSONResponse({"error": "代码片段不存在"}, status_code=404)

        snippet = dict(row)
        snippet['tags'] = json.loads(snippet['tags']) if snippet['tags'] else []

        # 增加使用次数
        cursor.execute("UPDATE snippets SET usage_count = usage_count + 1 WHERE id = ?", (snippet_id,))
        conn.commit()

        conn.close()

        return JSONResponse(snippet)
    except Exception as e:
        logger.exception(f"获取代码片段失败: {e}")
        return JSONResponse({"error": str(e)}, status_code=500)


@router.put("/{snippet_id}")
async def update_snippet(snippet_id: int, snippet: SnippetUpdate):
    """更新代码片段"""
    try:
        conn = get_db_connection()
        cursor = conn.cursor()

        # 检查是否存在
        cursor.execute("SELECT id FROM snippets WHERE id = ?", (snippet_id,))
        if not cursor.fetchone():
            conn.close()
            return JSONResponse({"error": "代码片段不存在"}, status_code=404)

        # 构建更新语句
        updates = []
        params = []

        if snippet.title is not None:
            updates.append("title = ?")
            params.append(snippet.title)
        if snippet.code is not None:
            updates.append("code = ?")
            params.append(snippet.code)
        if snippet.language is not None:
            updates.append("language = ?")
            params.append(snippet.language)
        if snippet.category is not None:
            updates.append("category = ?")
            params.append(snippet.category)
        if snippet.description is not None:
            updates.append("description = ?")
            params.append(snippet.description)
        if snippet.tags is not None:
            updates.append("tags = ?")
            params.append(json.dumps(snippet.tags))

        updates.append("updated_at = CURRENT_TIMESTAMP")
        params.append(snippet_id)

        query = f"UPDATE snippets SET {', '.join(updates)} WHERE id = ?"
        cursor.execute(query, params)

        conn.commit()
        conn.close()

        return JSONResponse({"message": "代码片段更新成功"})
    except Exception as e:
        logger.exception(f"更新代码片段失败: {e}")
        return JSONResponse({"error": str(e)}, status_code=500)


@router.delete("/{snippet_id}")
async def delete_snippet(snippet_id: int):
    """删除代码片段"""
    try:
        conn = get_db_connection()
        cursor = conn.cursor()

        cursor.execute("DELETE FROM snippets WHERE id = ?", (snippet_id,))

        if cursor.rowcount == 0:
            conn.close()
            return JSONResponse({"error": "代码片段不存在"}, status_code=404)

        conn.commit()
        conn.close()

        return JSONResponse({"message": "代码片段删除成功"})
    except Exception as e:
        logger.exception(f"删除代码片段失败: {e}")
        return JSONResponse({"error": str(e)}, status_code=500)


@router.get("/popular")
async def get_popular_snippets(limit: int = Query(10, ge=1, le=100)):
    """获取热门代码片段"""
    try:
        conn = get_db_connection()
        cursor = conn.cursor()

        cursor.execute("SELECT * FROM snippets ORDER BY usage_count DESC, updated_at DESC LIMIT ?", (limit,))
        rows = cursor.fetchall()

        snippets = []
        for row in rows:
            snippet = dict(row)
            snippet['tags'] = json.loads(snippet['tags']) if snippet['tags'] else []
            snippets.append(snippet)

        conn.close()

        return JSONResponse({"snippets": snippets})
    except Exception as e:
        logger.exception(f"获取热门代码片段失败: {e}")
        return JSONResponse({"error": str(e)}, status_code=500)


@router.get("/recent")
async def get_recent_snippets(limit: int = Query(10, ge=1, le=100)):
    """获取最近代码片段"""
    try:
        conn = get_db_connection()
        cursor = conn.cursor()

        cursor.execute("SELECT * FROM snippets ORDER BY updated_at DESC LIMIT ?", (limit,))
        rows = cursor.fetchall()

        snippets = []
        for row in rows:
            snippet = dict(row)
            snippet['tags'] = json.loads(snippet['tags']) if snippet['tags'] else []
            snippets.append(snippet)

        conn.close()

        return JSONResponse({"snippets": snippets})
    except Exception as e:
        logger.exception(f"获取最近代码片段失败: {e}")
        return JSONResponse({"error": str(e)}, status_code=500)


@router.post("/{snippet_id}/usage")
async def increment_snippet_usage(snippet_id: int):
    """增加代码片段使用次数"""
    try:
        conn = get_db_connection()
        cursor = conn.cursor()

        cursor.execute("UPDATE snippets SET usage_count = usage_count + 1 WHERE id = ?", (snippet_id,))

        if cursor.rowcount == 0:
            conn.close()
            return JSONResponse({"error": "代码片段不存在"}, status_code=404)

        conn.commit()
        conn.close()

        return JSONResponse({"message": "使用次数已更新"})
    except Exception as e:
        logger.exception(f"更新使用次数失败: {e}")
        return JSONResponse({"error": str(e)}, status_code=500)